Feature/web login

asincrono opened the pull request:
This component provides the login/logout functionality.
The authentication status is stored in the redux state by the presence of the  provided by the token we obtain from the server after a successful login. 
The error status is stored in the same way in the form of the error message.

asincrono committed:
Added dependencies from redux-thunk (6), redux-form and react-bootstrap.

asincrono committed:
Modal login form first approach.

Modal done with react-bootrap, form with redux-form.

asincrono committed:
Form field now done with reac-bootstrap form elements.

asincrono committed:
Adding actions. Cleaning form.

asincrono committed:
LoginModal and action are properly connected. Request done.

Added axios dependency to "catch" ERR_CONNECTION_REFUSED.

asincrono committed:
Main bar login "button", and username form field  tweaks.

asincrono committed:
Added action creators, logout button & minor button css tweaks.

Added the logout button (and funtionality) to the MainToolBar. Promoted to component as it will be aware of state change (authenticated) to show Log in / Log out option.
Minor tweaks to hide the outline in the Log in/out buttons.

asincrono committed:
Final version of Log in and Log out.

Login in and out combined in the element LogInAndOut.jsx.
MainToolBar.jsx back to being a function.
Set focus to unsermame input on modar login form.
Minor tweaks in the style of btn-as-navbar-link.

asincrono committed:
Almost there, but not quite yet .

Log in and  log out.
Added error message to log in modal through ErrorMessage component.
Added LOGIN/OUT_IN_PROGRESS states AC.

asincrono committed:
Just before first submission to code review keeping debug code.

asincrono committed:
Final version before code review.

Remove debug code.

asincrono committed:
Added user menu instead of logo ut button.

ssaavedra commented on xea-web/src/actions/authActions.js:
```diff
@@ -0,0 +1,138 @@
+import axios from 'axios'
+import { browserHistory } from 'react-router'
```
Para que necesitas aquí `browserHistory`?

asincrono replied:
É a forma na que o vin facer. Non coñecía hashHistory nin createMemoryHistory.
Cal crees ti que é a mellor forma de facelo?

ssaavedra replied:
Idealmente sería inyectar a dependencia do history apropiado dalgún xeito como teñen que facer os compoñentes Link e relacionados, que o collen dos nosos pais ou algo así, porque en esencia esto é como un Link, tería que funcionar no router que sexa que estea por encima. Un ollo ao código de Link debería de esclarecer algo. Estou revisando esto desde o móbil, senón xa cho contaba.

ssaavedra replied:
Se te fixas, en https://github.com/reactjs/react-router/blob/master/modules/Link.js tes como o fan eles.

Terías que ter acceso ao router, que llo terías que pasar ao AC.

¿Como?
O compoñente react correspondente debería de requerir en contexto un router, así:

``` javascript
import { routerShape } from 'react-router/PropTypes' // Esto creo que é accesible

component Blah extends Component {
  contextTypes: {
    router: routerShape
  },
```

E despois, unha vez tes o router, podes facer

``` javascript
this.context.router.push(location)
```

ssaavedra replied:
Pero esto se queres deixámolo como tarefa en taiga para refactor posterior, e ter xa esto merged.

Faille rebase, que esto non mergea en master ;)

ssaavedra commented on xea-web/src/actions/authActions.js:
```diff
@@ -0,0 +1,138 @@
+
+    axios(LOGOUT_URL, requestCfg)
+      .then(response => {
+        if (response.status === 204) {
```
Igual aquí pode ser máis razoable ser algo máis laxos, e deixar a comprobación como `response.status >=200 && response.status < 300`? Digo para non ter que axustar isto ante cambios pequenos da API.

asincrono replied:
Foi un pouco proba/error/preguntar. Estaría ben ter una lista coas respostas que se espera dea a API es así ter máis claro como implementar o cliente.
Póñome a facer o cambio.

ssaavedra commented on xea-web/src/actions/authActions.js:
```diff
@@ -0,0 +1,138 @@
+import axios from 'axios'
+import { browserHistory } from 'react-router'
```
O sea, o problema con browserHistory é que estás acoplando authActions ao tipo de historial que pides do navegador (porqué non hashHistory ou createMemoryHistory?)

Pero bueno, igual non é tan fácil de sacar de aquí...

ssaavedra commented on xea-web/src/actions/authActions.js:
```diff
@@ -0,0 +1,138 @@
+          })
+          // We redirect to the main page.
+          browserHistory.push('/main')
+        } else if (response.status === 401) {
```
Se tes un elseif, e non hai ningún outro else, eu diría de que este response.status ten que ser máis amplo, xa que se a API devolvese un 404 ou un 408 non cae por ningunha rama; así que isto debería de ser un condicional `response.status >= 400 && response.status < 500`.

E deberías de ter unha rama adicional

``` javascript
} else {
  throw new Error(response.status + ": " + response.statusText)
}
```

no caso dun error 500 ou algo parecido (que iso sería máis un LOGOUT_ERROR ca un LOGOUT_FAILURE).

asincrono replied:
entón o mellor sería que fixese que rexeitara os valores por riba de 500 co validateStatus:

``` javascript
//...
validateStatus: (status) => status < 500,
//...
```

Así todo o que sexa >= 500 xa o pilla o `catch` da promise e dispara un `LOGOUT_ERROR`.

ssaavedra replied:
Ah nice. Como podes ver, non me lin a documentación de axios ^_^U

ssaavedra commented on xea-web/src/components/login/LogInAndOut.jsx:
```diff
@@ -0,0 +1,117 @@
+    this.state = { showModal: false }
+
+    this.open = this.open.bind(this)
+    this.close = this.close.bind(this)
```
Nunca vira esta forma de facer as cousas. Non sería máis sencillo usar propiedades da clase de esta outra forma?

``` javascript
class LogInAndOut extends Component {
  open = () => { this.setState({showModal: true}) }

  // ...
}
```

asincrono replied:
Creo que non podes facer iso (estiven mirando e entendo que ES6 non permite propiedades de clase, so funcións/métodos)
Tería que ser no constructor:

``` javascript
class LogInAndOut extends Component {
  constructor(props) {
    this.open = () => {
      this.setState({ showModal: true })
      //...
    }
  }
```

Entón para saber o que pode facer una clase, en lugar de mirar ós seus métodos terías que mirar o constructor.
A verdade, non me gusta.

ssaavedra replied:
Agora mesmo ao constructor tres que mirar igual a ver se se fixo o bind a this, pero tes razón, non hai maneira que non teña defectos. Podemos ir coa túa :)

ssaavedra commented on xea-web/src/components/login/LogInAndOut.jsx:
```diff
@@ -0,0 +1,117 @@
+  }
+
+  close () {
+    // If login in progress or messageError we prevent closing modal.
```
Aquí parece que o comentario e o código non van sincronizados :-)

castrinho8 commented on xea-web/src/components/login/LoginModal.jsx:
```diff
@@ -0,0 +1,57 @@
+  }
+
+  handleFormSubmit (props) {
+    console.log('handleFormSubmit')
```
Si ya estás seguro de que va bien, yo borraría todos los console.log ;)

castrinho8 commented on xea-web/src/constants/urls.js:
```diff
@@ -0,0 +1,4 @@
+export const LOGIN_URL = 'http://localhost:8000/api/auth/jwt/get_token'
+export const VERIFY_URL = 'http://localhost:8000/api/auth/jwt/verfy_token'
+export const LOGOUT_URL = 'http://localhost:8000/api/auth/jwt/logout'
+export const LOGOUT_ALL_URL = 'http://localhost:8000/lapi/auth/jwt/ogout_all'
```
Esto igual puede quedar así de momento mientras no usamos otra cosa, pero aún así lo ideal sería tener las urls relativas en un archivo parecido a [este](https://github.com/vacmatch/vacmatch-mobile/blob/master/src/app/api/urls.js) y tener una función en ese fichero para construirlas con el "host", "port", etc que podrían estar metidos en un json que se podría leer, para que puedas seleccionar los datos en función de si estamos en DEV o en PRO:

`
{
  "production" : {
    "db": {
      "host": "192.168.1.101",
      "port": "5983"
    }
  },
  "development" : {
    "db": {
      "host": "localhost",
      "port": "5984"
    }
  },
  "_env": "production"
}
`

asincrono replied:
Si puedo miro de hacerlo como comentas mañana.

_¿Ese json de dónde se obtendría?_

castrinho8 replied:
Mismo en un directorio config en donde puedes poner tando el json como el fichero urls para todo el proyecto ;)

asincrono replied:
para hacer eso hubo que:

```
npm install --save json-loader
```

y luego en `webpack.config.dev.js`

``` javascript
module: {
    loaders: [
      //...
      { test: /\.json$/, exclude: /node_modules/, loader: 'json' },
      //...
```

asincrono committed:
Some fixes and improvemens after pull request comments.

Added support for loading json files (through json-loader)
New way of obtaining urls (urlconf.json).

ssaavedra commented on xea-web/src/constants/urls.js:
```diff
@@ -1,4 +1,40 @@
+  }
+
+  login () {
+    return `${this.baseUrl()}/api/auth/jwt/get_token`
```
Si en lugar de devolver este string, directamente devuelves axios(url, requestConf), consigues:
- Centralizar todas las llamadas a axios/fetch/lo que sea en un único sitio
- Que ese sitio no sean los action creators
- Que en lugar de llamarse URLs esto se llame API
- Que sea exactamente el acceso al API, una responsabilidad clara y desligada de las acciones
- Que se manejen ya las cosas como promises, que es lo que nos esperamos que ocurra anyway

¿Que opinas?

asincrono replied:
Me parece muy buena idea.
Pero con la suficientemente entidad como para que tenga su  propia tarea de Taiga y rama correspondiente. 
Si te parece bien dejamos esto así y empezamos a trabajar y definir API.

ssaavedra replied:
Me parece bien

asincrono committed:
Cleaning code 

very minor changes.

ssaavedra commented on xea-web/src/components/login/LoginForm.jsx:
```diff
@@ -141,6 +141,23 @@ const validate = ({ username, password }) => {
   return errors
 }
 
+// const validate = ({ username, password }) => {
```
El código comentado no sirve para mucho. Si esto no se usa, eliminarlo es el camino.

asincrono replied:
Maldito síndrome de Diógenes. Ya lo limpio.

ssaavedra commented on xea-web/src/components/login/LoginForm.jsx:
```diff
@@ -0,0 +1,172 @@
+    return 'success'
+  }
+
+  usernameNormalicer (value, previousValue) {
```
Typo: non debería de ser `usernameNormalizer`?

asincrono replied:
Ciertamente. A corregirlo.

ssaavedra commented on xea-web/src/components/login/LoginForm.jsx:
```diff
@@ -0,0 +1,172 @@
+    return value
+  }
+
+  renderField ({ name, placeholder, input, label, type,
```
Nit: estaría chulo se isto se puidese sacar do compoñente de login, pero podemos deixalo para o futuro.

asincrono replied:
O "normalizer"?
Igual poderían facerse "normalizers" para distintas cousas que se puidesen combinar é metelos todos nunha biblioteca. Se tal mellor esperar a ir atopando a necesidade de normalizar máis formularios?

ssaavedra replied:
Decía o renderField, realmente esto non podería ser un componente?

ssaavedra commented on xea-web/src/components/login/LoginForm.jsx:
```diff
@@ -0,0 +1,172 @@
+  return errors
+}
+
+// const validate = ({ username, password }) => {
```
Este código comentado sobra, xa hemos de ter bastante con manter o código que funciona.

asincrono committed:
Typo correction and comments removed.

ssaavedra commented:
LGTM

ssaavedra commented:
Haz el merge como squash si puedes, para que quede más limpio en la historia. En la historia del código poco importa ver cada commit de cada feature cuando lo que hay es una code review que va cambiando lo que hay en la rama.

Lo que sí es interesante es poner en el texto del commit "de que va" esta feature, para poder identificarla en el git log, exponiendo qué y por qué se hacen las cosas. Así que si puedes hacer squash y poner un mensaje más o menos largo en el que digas lo que se introduce aqui, sería fantástico.

ssaavedra commented:
Te volví a hacer el rebase porque mergeamos mi pull request. De nuevo, LGTM.
